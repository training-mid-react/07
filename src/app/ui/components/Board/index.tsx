import { FC } from 'react';
import './styles.scss';
import { Square } from '../Square';
import { ROWS } from '@core/constants';
import { OptionsColors } from '@core/types';

interface Props {
    board: OptionsColors[][];
    handleClick: (col: number) => void;
    fallingCells: string[];
}
export const Board: FC<Props> = ({ board, handleClick, fallingCells }) => (
    <div className="board">
        {board?.map((_, col) => (
            <div key={col} className="board__column">
                {Array.from({ length: ROWS }, (_, row) => (
                    <Square
                        key={`${row}-${col}`}
                        onClick={() => handleClick(col)}
                        value={board[row][col]}
                        isFalling={fallingCells.includes(`${row}-${col}`)}
                    />
                ))}
            </div>
        ))}
    </div>
);
