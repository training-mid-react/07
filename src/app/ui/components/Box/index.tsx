import { CELL } from "@/app/core/interfaces/game.interface";
import "./style.scss"

interface Props {
    value: CELL;
    onClick: () => void;
}

const Box = ({ value, onClick }: Props) => {
    return (
        <div data-testid="box" className="box" onClick={onClick}>
            {
                value && <div style={{backgroundColor: value.color}} className="box__chip"></div>
            }
        </div>
    );
}

export default Box;