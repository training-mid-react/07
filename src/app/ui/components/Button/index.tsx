import React from 'react';
import './style.scss';

interface ButtonProps {
  children: React.ReactNode;
  className?: string;
  onClick: () => void;
  disabled?: boolean;
}

const Button: React.FC<ButtonProps> = ({ children, onClick, disabled, className }) => {
  return (
    <button data-testid='button-component' className={className || 'custom-button'} onClick={onClick} disabled={disabled}>
      {children}
    </button>
  );
};

export default Button;
