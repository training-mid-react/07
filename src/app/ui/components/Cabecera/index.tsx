
import "./style.scss";

interface Props {
    currentPlayer: number
}

const Cabecera = ({ currentPlayer }: Props) => {
    return (
        <>
            <section className="cabecera">
                <div>
                    <h2>Jugador 1</h2>
                    <div className="ficha-cabecera ficha-cabecera--jugador1"></div>
                </div>
                <div>
                    <h2>Jugador 2</h2>
                    <div className="ficha-cabecera ficha-cabecera--jugador2"></div>
                </div>
            </section>
            <h3>Turno del jugador: {currentPlayer}</h3>
        </>

    )
}

export default Cabecera 