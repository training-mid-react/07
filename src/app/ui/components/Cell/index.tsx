import { FC } from 'react';
import { PLAYER } from '../../../../core/interfaces';
import './style.scss';

export interface CellProps {
  value: PLAYER | null;
}

export const Cell: FC<CellProps> = ({ value }) => {

  return (
    <>
    <div className={`cell cell__${value}`} />
    <div className='cell__frame'></div>
    </>
  );
};
