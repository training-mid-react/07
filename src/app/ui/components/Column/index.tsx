import './style.scss';
import { useDropCell } from '../../../core/hooks/useDropCell';

interface ColumnProps {
  columnIndex: number;
  dropPiece: (column: number) => void;
  columnData: number[];
}

function Column({ columnIndex, dropPiece, columnData }: ColumnProps) {
  const { handleColumnClick, droppingCell } = useDropCell({ columnIndex, dropPiece, columnData })
  return (
    <div className="column" onClick={handleColumnClick}>
      {columnData.map((cell, rowIndex) => {
        const isDropping = droppingCell === rowIndex;
        const cellClass = `cell ${cell === 1 ? 'cell--player1' : cell === 2 ? 'cell--player2' : ''} ${isDropping ? 'cell--dropping' : ''}`;
        return (
          <div
            key={rowIndex}
            className={cellClass}
          />
        );
      })}
    </div>
  );
};

export default Column;
