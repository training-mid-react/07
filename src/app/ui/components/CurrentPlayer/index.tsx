import './style.scss';

interface ICurrentPlayerProps {
  currentPlayer: number;
}

export const CurrentPlayer = ({ currentPlayer }: ICurrentPlayerProps) => {
  return (
    <div className="current-player">
      Es el turno del Jugador: 
      <span className={currentPlayer === 1 ? 'current-player__player1' : 'current-player__player2'}>
        {currentPlayer}
      </span>
    </div>
  );
};
