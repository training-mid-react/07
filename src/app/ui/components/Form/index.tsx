import { IPlayerForm } from '@core/interfaces';
import './styles.scss';
import { useCustomForm } from '@core/hooks';
import { FC } from 'react';

export const Form: FC<{
    onSubmit: (data: IPlayerForm) => void;
}> = ({ onSubmit }) => {
    const { selectOptions, register, handleSubmit, errors } = useCustomForm();

    return (
        <form className="form" onSubmit={handleSubmit(onSubmit)}>
            <h2>Configuraci√≥n de Jugadores</h2>

            <h3>Jugador 1</h3>
            <label>
                Nombre:
                <input
                    role="input"
                    type="text"
                    {...register('player1.name', {
                        required: 'Nombre es requerido',
                    })}
                    placeholder="Ingresa un nombre"
                />
                {errors.player1?.name && (
                    <span style={{ color: 'red' }}>
                        {errors.player1.name.message}
                    </span>
                )}
            </label>
            <label>
                Color:
                <select
                    {...register('player1.color', {
                        required: 'Selecciona un color',
                    })}
                >
                    {selectOptions.map((o) => (
                        <option key={o.value} value={o.value}>
                            {o.label}
                        </option>
                    ))}
                </select>
                {errors.player1?.color && (
                    <span style={{ color: 'red' }}>
                        {errors.player1.color.message}
                    </span>
                )}
            </label>

            <h3>Jugador 2</h3>
            <label>
                Nombre:
                <input
                    type="text"
                    {...register('player2.name', {
                        required: 'Nombre es requerido',
                    })}
                    placeholder="Ingresa un nombre"
                />
                {errors.player2?.name && (
                    <span style={{ color: 'red' }}>
                        {errors.player2.name.message}
                    </span>
                )}
            </label>
            <label>
                Color:
                <select
                    {...register('player2.color', {
                        required: 'Selecciona un color',
                    })}
                >
                    {selectOptions.map((o) => (
                        <option key={o.value} value={o.value}>
                            {o.label}
                        </option>
                    ))}
                </select>
                {errors.player2?.color && (
                    <span style={{ color: 'red' }}>
                        {errors.player2.color.message}
                    </span>
                )}
            </label>

            <button>Iniciar Juego</button>
        </form>
    );
};
