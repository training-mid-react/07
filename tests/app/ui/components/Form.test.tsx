import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { Form } from '../../../../src/app/ui/components/Form';
import { describe, expect, test, vi } from 'vitest';
import React from 'react';
import '@testing-library/jest-dom';

vi.mock('@core/hooks', () => ({
    useform: () => ({
        selectOptions: [
            { label: 'Rojo', value: 'red' },
            { label: 'Azul', value: 'blue' },
        ],
        register: vi.fn(),
        handleSubmit: (callback: any) => callback,
        errors: {},
        setValue: vi.fn(),
        getValues: () => ({
            player1: { name: 'Player 1', color: 'blue' },
            player2: { name: 'Player 2', color: 'rojo' },
        }),
    }),
}));

describe('Form', () => {
    const mockOnSubmit = vi.fn((data) => {
        Promise.resolve();
    });

    test('Match snapshoot', () => {
        const { asFragment } = render(<Form onSubmit={mockOnSubmit} />);
        expect(asFragment()).toMatchSnapshot();
    });

    test('you should render the form correctly', () => {
        render(<Form onSubmit={mockOnSubmit} />);

        const title = screen.getByText('ConfiguraciÃ³n de Jugadores');
        expect(title).toBeInTheDocument();

        const inputs = screen.getAllByPlaceholderText('Ingresa un nombre');

        inputs.forEach((input) => {
            expect(input).toBeInTheDocument();
        });
        const player1 = screen.getByText('Jugador 1');
        const player2 = screen.getByText('Jugador 2');

        expect(player1).toBeInTheDocument();
        expect(player2).toBeInTheDocument();

        const names = screen.getAllByLabelText(/Nombre/);
        names.forEach((name) => {
            expect(name).toBeInTheDocument();
        });

        const labels = screen.getAllByLabelText(/Color/);
        labels.forEach((label) => {
            expect(label).toBeInTheDocument();
        });

        expect(screen.getByText('Iniciar Juego')).toBeInTheDocument();
    });

    test('function is not called if data is not completed', async () => {
        const { getByRole } = render(<Form onSubmit={mockOnSubmit} />);
        const button = getByRole('button', { name: /Iniciar Juego/i });
        await waitFor(() => fireEvent.click(button));

        const alert = await screen.findAllByText('Nombre es requerido');
        alert.forEach((a) => expect(a).toBeInTheDocument());

        expect(mockOnSubmit).not.toHaveBeenCalled();
    });

    test('function is not called if data is wrong', async () => {
        const { container, getAllByLabelText } = render(
            <Form onSubmit={mockOnSubmit} />
        );

        const selectColor = getAllByLabelText('Color:');

        for (const [_, select] of selectColor.entries())
            await waitFor(() => {
                fireEvent.change(select, {
                    target: { value: 'blue' },
                });
                fireEvent.blur(select);
            });

        expect(container.innerHTML).toMatch(
            'Los jugadores no pueden compartir el mismo color'
        );
    });

    test('function is called if data is completed', async () => {
        render(<Form onSubmit={mockOnSubmit} />);

        const { getAllByLabelText, getByRole } = screen;

        const inputsName = getAllByLabelText('Nombre:');
        const selectColor = getAllByLabelText('Color:');
        for (const [index, input] of inputsName.entries()) {
            await waitFor(() =>
                fireEvent.change(input, {
                    target: { value: `Player ${index + 1}` },
                })
            );
        }
        for (const [index, select] of selectColor.entries()) {
            await waitFor(() =>
                fireEvent.change(select, {
                    target: { value: `${index === 0 ? 'blue' : 'red'}` },
                })
            );
        }

        const submitButton = getByRole('button', { name: /Iniciar Juego/i });
        await waitFor(() => fireEvent.click(submitButton));

        expect(mockOnSubmit).toHaveBeenCalledWith(
            {
                player1: { name: 'Player 1', color: 'blue' },
                player2: { name: 'Player 2', color: 'red' },
            },
            expect.objectContaining({
                type: 'submit',
            })
        );
    });
});
