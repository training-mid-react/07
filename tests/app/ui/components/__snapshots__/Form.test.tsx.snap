// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Form > Match snapshoot 1`] = `
<DocumentFragment>
  <form
    class="form"
  >
    <h2>
      Configuraci√≥n de Jugadores
    </h2>
    <h3>
      Jugador 1
    </h3>
    <label>
      Nombre:
      <input
        name="player1.name"
        placeholder="Ingresa un nombre"
        role="input"
        type="text"
      />
    </label>
    <label>
      Color:
      <select
        name="player1.color"
      >
        <option
          value=""
        >
          Selecciona un color
        </option>
        <option
          value="red"
        >
          Rojo
        </option>
        <option
          value="blue"
        >
          Azul
        </option>
      </select>
    </label>
    <h3>
      Jugador 2
    </h3>
    <label>
      Nombre:
      <input
        name="player2.name"
        placeholder="Ingresa un nombre"
        type="text"
      />
    </label>
    <label>
      Color:
      <select
        name="player2.color"
      >
        <option
          value=""
        >
          Selecciona un color
        </option>
        <option
          value="red"
        >
          Rojo
        </option>
        <option
          value="blue"
        >
          Azul
        </option>
      </select>
    </label>
    <button>
      Iniciar Juego
    </button>
  </form>
</DocumentFragment>
`;
